### PR TITLE
Update readme for hreader response

### PR DESCRIPTION
**Pull request by KhushbooMer**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -17,6 +17,7 @@ Table of contents -
  * [Installation](#installation)
  * [Initialization](#initialization)
```
Comment (imagekitio) [inline README.md:19]: Should we mention the solution suggested in below issue in Initialization section?
https://github.com/imagekit-developer/imagekit-java/issues/10

File: README.md [hunk 2]
```diff
@@ -94,6 +95,36 @@ class App {
 You can use this Java SDK for 3 different kinds of methods - URL generation, file upload, and file management.
```
Comment (imagekitio) [inline README.md]: Use bullet points.

File: README.md [hunk 3]
```diff
@@ -94,6 +95,36 @@ class App {
...
 The usage of the SDK has been explained below.
 
+## Versioning
```
Comment (imagekitio) [inline README.md]: Should the heading name be "Change log" or `Upgrading to 2.x.x"?

**Review by imagekitio — CHANGES_REQUESTED** [review]
Add support for `raw` transformation parameter so that user can pass any transformation string as it is.
Ruby - https://github.com/imagekit-developer/imagekit-ruby
image

Also add a test case like this - https://github.com/imagekit-developer/imagekit-ruby/blob/master/test/imagekit/url_test.rb#L142

File: imagekit-sdk/build.gradle [hunk 4]
```diff
@@ -13,7 +13,12 @@ repositories {
 }
 
 dependencies {
-    compile fileTree(include: ['*.jar'], dir: '../libs')
+    compile 'commons-codec:commons-codec:1.13'
+    implementation 'com.squareup.retrofit2:converter-gson:2.3.0'
+    implementation 'com.google.code.gson:gson:2.7'
+    implementation 'com.squareup.okhttp3:okhttp:3.10.0'
```
Comment (imagekitio) [inline imagekit-sdk/build.gradle:23]: What happens if someone is already using another version of okhttp3?

**Review by imagekitio — CHANGES_REQUESTED** [review]

**Review by imagekitio — CHANGES_REQUESTED** [review]
let's get rid of `ik-sdk-version` parameter in generated URL. Update the test cases accordingly.

File: imagekit-sdk/src/test/java/io/imagekit/sdk/ImageKitTest.java [hunk 5]
```diff
@@ -2,629 +2,1890 @@
...
+	public void getUrl_with_overriding_urlEndpoint() {
+		List<Map<String, String>> transformation = new ArrayList<Map<String, String>>();
+		Map<String, String> scale = new HashMap<>();
+		scale.put("height", "600");
+		scale.put("width", "400");
+		transformation.add(scale);
+
+		Map<String, Object> options = new HashMap<>();
+		options.put("path", "/default-image.jpg");
+		options.put("urlEndpoint", "https://ik.imagekit.io/your_override_imagekit_id");
```
Comment (imagekitio) [inline imagekit-sdk/src/test/java/io/imagekit/sdk/ImageKitTest.java]: Replace `https://ik.imagekit.io/your_override_imagekit_id` with `https://ik.imagekit.io/your_imagekit_id/different-url-endpoint-prefix`

File: imagekit-sdk/src/test/java/io/imagekit/sdk/ImageKitTest.java [hunk 6]
```diff
@@ -2,629 +2,1890 @@
...
+
+	@Test
+	public void getUrl_with_overriding_urlEndpoint_double_slash_tests() {
+		List<Map<String, String>> transformation = new ArrayList<Map<String, String>>();
+		Map<String, String> scale = new HashMap<>();
+		scale.put("height", "600");
+		scale.put("width", "400");
+		transformation.add(scale);
+
+		Map<String, Object> options = new HashMap<>();
```
Comment (imagekitio) [inline imagekit-sdk/src/test/java/io/imagekit/sdk/ImageKitTest.java]: Test case is missing a double slash actually. I see the original test case was incorrect too. Change the value of `https://ik.imagekit.io/your_override_imagekit_id` to `https://ik.imagekit.io/your_imagekit_id/`

File: imagekit-sdk/src/test/java/io/imagekit/sdk/ImageKitTest.java [hunk 7]
```diff
@@ -2,629 +2,1890 @@
...
+		options.put("transformation", transformation);
+
+		String url = SUT.getUrl(options);
+		assertThat("https://ik.imagekit.io/your_imagekit_id/endpoint/default-image.jpg?ik-sdk-version="
+				+ Version.VERSION_CODE + "&srcParam=srcParamValue&v=123&z=234&tr=w-400,h-600", is(url));
+	}
+
+	@Test
+	public void getUrl_src_with_query_params_but_transformationPosition_is_path() {
+		Map<String, String> queryParam = new HashMap<>();
```
Comment (imagekitio) [inline imagekit-sdk/src/test/java/io/imagekit/sdk/ImageKitTest.java]: Include a gigantic test case that coveres all the parameter including `raw` like this https://github.com/imagekit-developer/imagekit-ruby/blob/master/test/imagekit/url_test.rb#L142

Notice how value of `overlay_image` and `default_image` has been changed such that `/` is replaced with `@@`.
Docs - https://docs.imagekit.io/features/image-transformations/overlay#using-overlay-image-that-are-stored-in-a-nested-folder

File: imagekit-sdk/src/test/java/io/imagekit/sdk/ImageKitTest.java [hunk 8]
```diff
@@ -2,629 +2,1890 @@
...
+		innerObject2.addProperty("name", "google-auto-tagging");
+		innerObject2.addProperty("minConfidence", 10);
+		innerObject2.addProperty("maxTags", 5);
+		JsonArray jsonArray = new JsonArray();
+		jsonArray.add(innerObject1);
+		jsonArray.add(innerObject2);
+		fileCreateRequest.setExtensions(jsonArray);
+		fileCreateRequest.setWebhookUrl("https://webhook.site/c78d617f-33bc-40d9-9e61-608999721e2e");
+		fileCreateRequest.setUseUniqueFileName(false);
+		fileCreateRequest.setPrivateFile(false);
```
Comment (imagekitio) [inline imagekit-sdk/src/test/java/io/imagekit/sdk/ImageKitTest.java]: Have some `true` values as well. Mix and match.

File: imagekit-sdk/src/test/java/io/imagekit/sdk/ImageKitTest.java [hunk 9]
```diff
@@ -2,629 +2,1890 @@
...
+	public void imageKit_updateDetails_expected_404()
+			throws IOException, InterruptedException, BadRequestException, InternalServerException, UnknownException,
+			ForbiddenException, TooManyRequestsException, UnauthorizedException {
+
+		FileUpdateRequest fileUpdateRequest = new FileUpdateRequest("62a9c3ccd875ec6fd658c804");
+		MockWebServer server = new MockWebServer();
+		String responseJson = "{\n" + "    \"message\": \"The requested file does not exist.\",\n"
+				+ "    \"help\": \"For support kindly contact us at support@imagekit.io .\"\n" + "}";
+		server.enqueue(new MockResponse().setResponseCode(404).setBody(responseJson));
+		server.start();
```
Comment (imagekitio) [inline imagekit-sdk/src/test/java/io/imagekit/sdk/ImageKitTest.java]: What are we asserting in this test case?

File: imagekit-sdk/src/test/java/io/imagekit/sdk/ImageKitTest.java [hunk 10]
```diff
@@ -2,629 +2,1890 @@
...
+		RestClient.API_BASE_URL = server.url("/").toString();
+		SUT.updateFileDetail(fileUpdateRequest);
+		server.takeRequest();
+	}
+
```
Comment (imagekitio) [inline imagekit-sdk/src/test/java/io/imagekit/sdk/ImageKitTest.java]: It seems that List file API test cases are missing. Also take care of using passing `tags` as an array and SDK converting it to comma separated values properly. Also same for boolean fields conversion.

**Review by imagekitio — CHANGES_REQUESTED** [review]

File: imagekit-sdk/src/test/java/io/imagekit/sdk/ImageKitTest.java (hunk 9 above)
Comment (amanforindia) [reply imagekit-sdk/src/test/java/io/imagekit/sdk/ImageKitTest.java]: @imagekitio the assert is on line 586 (expected = NotFoundException.class)

**Review by imagekitio — CHANGES_REQUESTED** [review]
Error handling test case missing for `.upload()`. We only have one successful test case. We need to have basic required parameter missing test case where SDK throws error itself if `file` and `fileName` is missing. Also we need to have test case for when backend (mock in test case) returns a non 2xx response along with an error body as per docs.

**Review by imagekitio — CHANGES_REQUESTED** [review]
Same for `getFileList`, only one successful test case is present. No edge cases. We need to have basic error handling test cases for all methods.

**Review by imagekitio — CHANGES_REQUESTED** [review]
Split this huge file `imagekit-sdk/src/test/java/io/imagekit/sdk/ImageKitTest.java` into separate small files so that it is easy to review test suit. Create separate file names e.g. `upload.java`, `manage->file.java` etc. See node.js structure
image

File: imagekit-sdk/src/test/java/io/imagekit/sdk/ImageKitTest.java [hunk 11]
```diff
@@ -2,629 +2,1890 @@
...
+				+ "        \"url\": \"https://ik.imagekit.io/zv3rkhsym/default-image.jpg\",\n"
+				+ "        \"thumbnail\": \"https://ik.imagekit.io/zv3rkhsym/tr:n-ik_ml_thumbnail/default-image.jpg\",\n"
+				+ "        \"fileType\": \"image\",\n" + "        \"filePath\": \"/default-image.jpg\",\n"
+				+ "        \"height\": 1000,\n" + "        \"width\": 1000,\n" + "        \"size\": 147022,\n"
+				+ "        \"hasAlpha\": false,\n" + "        \"mime\": \"image/jpeg\"\n" + "    }\n" + "]";
+		server.enqueue(new MockResponse().setBody(responseJson));
+		server.start();
+		RestClient.API_BASE_URL = server.url("/").toString();
+		Map<String, String> options = new HashMap<>();
+		options.put("limit", "" + 1);
```
Comment (imagekitio) [inline imagekit-sdk/src/test/java/io/imagekit/sdk/ImageKitTest.java]: Other parameters are not included. Also if the user passes `tags` as an array, will it be converted into a comma-separated string by SDK. All this needs to be asserted in test cases.

File: imagekit-sdk/src/test/java/io/imagekit/sdk/ImageKitTest.java [hunk 12]
```diff
@@ -2,629 +2,1890 @@
...
+			throws IOException, InterruptedException, BadRequestException, InternalServerException, UnknownException,
+			ForbiddenException, TooManyRequestsException, UnauthorizedException {
+
+		MockWebServer server = new MockWebServer();
+		String responseJson = "{\n" + "    \"message\": \"The requested file does not exist.\",\n"
+				+ "    \"help\": \"For support kindly contact us at support@imagekit.io .\"\n" + "}";
+		server.enqueue(new MockResponse().setResponseCode(404).setBody(responseJson));
+		server.start();
+		RestClient.API_BASE_URL = server.url("/").toString();
+		SUT.getFileDetail("629f3de17eb0fe4053615450");
```
Comment (imagekitio) [inline imagekit-sdk/src/test/java/io/imagekit/sdk/ImageKitTest.java]: What are we asserting in this test case?

File: imagekit-sdk/src/test/java/io/imagekit/sdk/ImageKitTest.java [hunk 13]
```diff
@@ -2,629 +2,1890 @@
...
+		RecordedRequest request = server.takeRequest();
+
+		assertEquals("application/json", request.getHeader("Content-Type"));
+		assertEquals("GET /v1/bulkJobs/629f44ac7eb0fe8173622d4b HTTP/1.1", request.getRequestLine());
+		assertEquals(RestClient.API_BASE_URL.concat("v1/bulkJobs/629f44ac7eb0fe8173622d4b"),
+				request.getRequestUrl().toString());
+	}
+
+	@Test(expected = NotFoundException.class)
+	public void getFileVersions_404_Expected()
```
Comment (imagekitio) [inline imagekit-sdk/src/test/java/io/imagekit/sdk/ImageKitTest.java]: Missing test case for `updateDetail`

File: imagekit-sdk/src/test/java/io/imagekit/sdk/ImageKitTest.java [hunk 14]
```diff
@@ -2,629 +2,1890 @@
...
+			throws InterruptedException, IOException, NotFoundException, BadRequestException, InternalServerException,
+			UnknownException, ForbiddenException, TooManyRequestsException, UnauthorizedException {
+
```
Comment (imagekitio) [inline imagekit-sdk/src/test/java/io/imagekit/sdk/ImageKitTest.java]: Missing test case for `manageTags`

Also , shouldn't we introduce a new method instead of overloading same `manageTags` function like below.
```java
// Remove tags example
List<String> fileIds = new ArrayList<>();
fileIds.add("FileId");
List<String> tags = new ArrayList<>();
tags.add("tag-to-remove-1");
tags.add("tag-to-remove-2");
ResultTags resultTags=ImageKit.getInstance().manageTags(new TagsRequest(fileIds, tags), "removeTags");
System.out.println("======FINAL RESULT=======");
System.out.println(resultTags);
System.out.println("Raw Response:");
System.out.println(resultTags.getResponseMetaData().getRaw());
System.out.println("Map Response:");
System.out.println(resultTags.getResponseMetaData().getMap());
```

File: imagekit-sdk/src/test/java/io/imagekit/sdk/ImageKitTest.java [hunk 15]
```diff
@@ -2,629 +2,1890 @@
...
+	@Test
+	public void renameFile_successExpected() throws InterruptedException, IOException, ConflictException,
+			PartialSuccessException, NotFoundException, BadRequestException, InternalServerException, UnknownException,
+			ForbiddenException, TooManyRequestsException, UnauthorizedException {
+
+		RenameFileRequest renameFileRequest = new RenameFileRequest();
+		renameFileRequest.setFilePath("/car_false.jpeg");
+		renameFileRequest.setNewFileName("new_car.jpeg");
+		renameFileRequest.setPurgeCache(true);
```
Comment (imagekitio) [inline imagekit-sdk/src/test/java/io/imagekit/sdk/ImageKitTest.java]: Include test case for `false` condition as well `renameFileRequest.setPurgeCache(false);`

**Review by imagekitio — CHANGES_REQUESTED** [review]
One API missing https://docs.imagekit.io/api-reference/media-api/restore-file-version. Update readme and test cases as well.

File: imagekit-sdk/src/test/java/io/imagekit/sdk/ImageKitTest.java [hunk 16]
```diff
@@ -2,629 +2,1890 @@
...
+
+	@Test(expected = NotFoundException.class)
+	public void copyFolder_404_Expected()
+			throws InterruptedException, IOException, NotFoundException, BadRequestException, InternalServerException,
+			UnknownException, ForbiddenException, TooManyRequestsException, UnauthorizedException {
+
+		CopyFolderRequest copyFolderRequest = new CopyFolderRequest();
+		copyFolderRequest.setSourceFolderPath("/testFolder");
+		copyFolderRequest.setDestinationPath("/test");
+		copyFolderRequest.setIncludeVersions(true);
```
Comment (imagekitio) [inline imagekit-sdk/src/test/java/io/imagekit/sdk/ImageKitTest.java]: Include `false` case as well.

File: imagekit-sdk/src/test/java/io/imagekit/sdk/ImageKitTest.java [hunk 17]
```diff
@@ -2,629 +2,1890 @@
...
+		server.enqueue(new MockResponse().setResponseCode(404)
+				.setBody("{\n" + "    \"message\": \"No files & folder found at sourceFolderPath /testFolder\",\n"
+						+ "    \"help\": \"For support kindly contact us at support@imagekit.io .\",\n"
+						+ "    \"reason\": \"NO_FILES_FOLDER\"\n" + "}"));
+		server.start();
+		RestClient.API_BASE_URL = server.url("/").toString();
+		SUT.copyFolder(copyFolderRequest);
+		server.takeRequest();
+	}
+
```
Comment (imagekitio) [inline imagekit-sdk/src/test/java/io/imagekit/sdk/ImageKitTest.java]: `.getFileMetadata` test cases missing.

**Review by imagekitio — CHANGES_REQUESTED** [review]

File: imagekit-sdk/src/test/java/io/imagekit/sdk/ImageKitTest.java [hunk 18]
```diff
@@ -2,629 +2,1890 @@
...
+		server.enqueue(new MockResponse().setBody(responseJson));
+		server.start();
+		RestClient.API_BASE_URL = server.url("/").toString();
+		SUT.getRemoteFileMetadata("https://ik.imagekit.io/zv3rkhsym/default-image.jpg");
+
+		RecordedRequest request = server.takeRequest();
+		String utf8RequestBody = request.getBody().readUtf8();
+		assertEquals("", utf8RequestBody);
+		assertEquals("application/json", request.getHeader("Content-Type"));
+		assertEquals("GET /v1/metadata?url=https://ik.imagekit.io/zv3rkhsym/default-image.jpg HTTP/1.1",
```
Comment (imagekitio) [inline imagekit-sdk/src/test/java/io/imagekit/sdk/ImageKitTest.java]: Do we need to handle any other cases where url is missing?

File: README.md [hunk 19]
```diff
@@ -394,9 +444,9 @@ ResultMetaData result=ImageKit.getInstance().getRemoteFileMetadata(url);
 System.out.println("======FINAL RESULT=======");
 System.out.println(result);
 System.out.println("Raw Response:");
-System.out.println(result.getRaw());
+System.out.println(result.getResponseMetaData().getRaw());
 System.out.println("Map Response:");
-System.out.println(result.getMap());
+System.out.println(result.getResponseMetaData().getMap());
 ```
```
Comment (imagekitio) [inline README.md:479]: The order of APIs in readme.md should match that of checklit in Notion document or what is there on docs.imagekit.io. Currently it seems that readme.md was updated as soon as that API implementation was done and it is all mixed up and out of any order and structure.

**Review by imagekitio — CHANGES_REQUESTED** [review]

File: imagekit-sdk/build.gradle (hunk 4 above)
Comment (amanforindia) [reply imagekit-sdk/build.gradle:23]: gradle picks the higher version:
https://docs.gradle.org/current/userguide/dependency_resolution.html#sec:version-conflict

File: imagekit-sdk/src/test/java/io/imagekit/sdk/ImageKitTest.java (hunk 12 above)
Comment (amanforindia) [reply imagekit-sdk/src/test/java/io/imagekit/sdk/ImageKitTest.java]: @imagekitio the assert is on line with (expected = NotFoundException.class)

File: imagekit-sdk/src/test/java/io/imagekit/sdk/ImageKitTest.java (hunk 13 above)
Comment (amanforindia) [reply imagekit-sdk/src/test/java/io/imagekit/sdk/ImageKitTest.java]: They are already present. Check these functions: imageKit_updateDetails_*

File: imagekit-sdk/src/test/java/io/imagekit/sdk/ImageKitTest.java (hunk 14 above)
Comment (amanforindia) [reply imagekit-sdk/src/test/java/io/imagekit/sdk/ImageKitTest.java]: I think it is not needed. `manageTags` is a helper function for the add and remove tags functions.

File: imagekit-sdk/src/test/java/io/imagekit/sdk/ImageKitTest.java (hunk 18 above)
Comment (amanforindia) [reply imagekit-sdk/src/test/java/io/imagekit/sdk/ImageKitTest.java]: It is already covered in the test case for 400 response code for this metadata endpoint.

**Review by imagekitio — APPROVED** [review]
